### PR TITLE
Periodically refresh salesforce auth token, rather than on demand

### PR DESCRIPTION
**Pull request by laurence-bird**

### CODE REVIEW

File: src/test/scala/com/ovoenergy/comms/profiles/salesforce/AuthStreamSpec.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  // TODO: use TestContext once we update to latest cats-effect/fs2 to avoid delays
+  "ApiStreamSpec" when {
+    "calls to authenticate are successfuil" should {
+      "emit a new value every period" in {
+        val auth = new SalesforceAuthClient[IO] {
+          override def authenticate(implicit F: Sync[IO]): IO[AuthenticationDetails] =
+            IO(AuthenticationDetails(UUID.randomUUID().toString, "test.com"))
+        }
+
+        val resultStream = Scheduler.apply[IO](1).flatMap { sch: Scheduler =>
```
Comment (filosganga) [inline src/test/scala/com/ovoenergy/comms/profiles/salesforce/AuthStreamSpec.scala]: you could do `implicit sch: Scheduler =>`

File: src/test/scala/com/ovoenergy/comms/profiles/salesforce/AuthStreamSpec.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          }
+        }
+
+        val res = resultStream.compile.toVector.unsafeRunSync()
+        res.size shouldBe 10
+        res.map(_.access_token).toSet.size shouldBe 10
+      }
+    }
+
+    "calls to authenticate which are unsuccessful" should {
```
Comment (filosganga) [inline src/test/scala/com/ovoenergy/comms/profiles/salesforce/AuthStreamSpec.scala]: I would add a test scenario like this:
 - success
 - failure
 - success

File: src/servicetest/scala/com/ovoenergy/comms/profiles/ProfilesKit.scala [hunk 3]
```diff
@@ -62,7 +62,7 @@ trait ProfilesKit extends DockerTestKit with DockerHostIpProvider with DockerLog
     )
     .withExposedPorts(profilesPort)
     .withVolumeBindings(List(HostConfig.Bind.from(s"""${sys.props("user.home")}/.aws""").to("/sbin/.aws").build()))
-    .withReadyChecker(HttpResponseCode(port = profilesPort, path = "/healthcheck", code = 200).looped(10, 5.seconds))
+    .withReadyChecker(LogLineContains("Kafka version"))
```
[the remark this answers is not in the document]
Comment (filosganga) [reply src/servicetest/scala/com/ovoenergy/comms/profiles/ProfilesKit.scala]: I see, it makes sense then.

File: src/main/scala/com/ovoenergy/comms/profiles/Config.scala [hunk 4]
```diff
@@ -22,6 +22,7 @@ import com.ovoenergy.comms.profiles.domain.{CustomerId, Realm}
 import com.ovoenergy.comms.profiles.http.ApiKeyAuthenticationMiddleware.ApiKey
 import com.ovoenergy.fs2.kafka.ConsumerSettings
 import com.ovoenergy.kafka.serialization.avro.SchemaRegistryClientSettings
+import fs2.internal
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/profiles/Config.scala]: ???

File: src/main/scala/com/ovoenergy/comms/profiles/Main.scala [hunk 5]
```diff
@@ -35,45 +34,44 @@ object Main extends StreamApp[IO] with AvroCodec {
...
-
-            val customerLookup: CustomerId => IO[Option[Customer]] = {
-              case CanaryCustomerId => canaryCustomer.some.pure[IO]
-
-              case CustomerId(Realm.Legacy, value) =>
-                Salesforce.fetchCustomerFromSf[IO](Salesforce.makeAuthenticatedRequest(config.salesforce), value)
-
-              case customerId: CustomerId =>
-                customerRepository.fetchCustomerById(customerId)
+            val httpClient = HttpClient.apply _
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/profiles/Main.scala]: a bit confused by this

File: src/main/scala/com/ovoenergy/comms/profiles/salesforce/SalesforceAuthClient.scala [hunk 6]
```diff
@@ -1,101 +1,74 @@
...
-
-    def retryIfUnauthorised(response: Response) = {
-      if (response.code == 401) {
-        reauthenticate().right.flatMap { retryAuthDetails =>
-          buildAndExecuteRequest(retryAuthDetails.instance_url, retryAuthDetails.access_token)
-        }
-      } else Right(response)
-    }
+trait SalesforceAuthClient[F[_]] {
+  def authenticate(implicit F: Sync[F]): F[AuthenticationDetails]
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/profiles/salesforce/SalesforceAuthClient.scala]: No need to put a constraint on the method here

File: src/main/scala/com/ovoenergy/comms/profiles/salesforce/SalesforceAuthClient.scala [hunk 7]
```diff
@@ -1,101 +1,74 @@
...
+        .addQueryParameter("client_id", authConfig.clientId.value)
+        .addQueryParameter("client_secret", authConfig.clientSecret.value)
+        .addQueryParameter("username", authConfig.username)
+        .addQueryParameter("password", authConfig.password.value)
+        .build()
+
+      import scala.util.control.NonFatal
+
+      F.delay(logger.info(s"Requesting: ${url.toString}")) >> F
+        .delay(htpClient(new Request.Builder().post(requestBody).url(url).build()))
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/profiles/salesforce/SalesforceAuthClient.scala]: `httpClient`

**Pull request merged**